### PR TITLE
e2e: adds kind test for deployment manifests

### PR DESCRIPTION
**Pull request by nilekhc**
Signed-off-by: Nilekh Chaudhari <1626598+nilekhc@users.noreply.github.com>

**Reason for Change**:

This PR adds test for driver and provider resources. (Non helm install tests)

**Requirements**

- [x] squashed commits
- [ ] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

fixes: #196 

**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**If "Yes," did you notify that project's maintainers and provide attribution?**

**Special Notes for Reviewers**:

### CODE REVIEW

File: .pipelines/templates/e2e-test-kind.yaml [hunk 1]
```diff
@@ -1,54 +1,59 @@
 parameters:
   - name: driverWriteSecrets
     type: object
+  - name: helmTests
+    type: object
 
 jobs:
   - ${{ each driverWriteSecret in parameters.driverWriteSecrets }}:
```
Comment (aramase) [inline .pipelines/templates/e2e-test-kind.yaml:8]: `--driver-write-secrets` is only applicable with helm. For the deploy manifests the default is `true` and we aren't overriding it later. So maybe we can just have the yaml tests as a separate job here instead of adding to the matrix?

File: deployment/provider-azure-installer.yaml [hunk 2]
```diff
@@ -6,11 +6,14 @@ metadata:
 apiVersion: apps/v1
 kind: DaemonSet
 metadata:
+  name: csi-csi-secrets-store-provider-azure
```
Comment (aramase) [inline deployment/provider-azure-installer.yaml:9]: Was this change intended? All yaml changes only in `manifest_staging/deployment`

File: test/e2e/e2e_suite_test.go [hunk 3]
```diff
@@ -41,27 +42,31 @@ var _ = BeforeSuite(func() {
...
-			helm.Install(helm.InstallInput{
-				Config: config,
-			})
-		} else if config.IsBackwardCompatibilityTest {
-			//We upgrade only if we are running backward compatibility tests.
-			By(fmt.Sprintf("Upgrading Secrets Store CSI Driver and Azure Key Vault Provider via Helm to New Version from - %s.", config.HelmChartDir))
-			helm.Upgrade(helm.UpgradeInput{
-				Config: config,
-			})
+		if config.IsHelmTest {
```
Comment (aramase) [inline test/e2e/e2e_suite_test.go:52]: This is too many nested if else. I think we can return early on some of the scenarios and reduce the nesting.

1. If it's soak, return.
2. If it's not helm test, deploy with manifest and return.
3. If none of the above perform the checks for helm release exists.

File: test/e2e/framework/deploy/deploy.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		"secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml",
+	}
+
+	providerResources = []string{
+		"provider-azure-installer.yaml",
+		"provider-azure-installer-windows.yaml",
+	}
+)
+
+func InstallManifest(kubeconfigPath string) {
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:37]: Let's add a comment here

File: test/e2e/framework/deploy/deploy.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	for _, resource := range driverResources {
+		exec.KubectlApply(kubeconfigPath, framework.NamespaceKubeSystem, []string{"-f", fmt.Sprintf("%s/%s", driverResourcePath, resource)})
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:39]: We should check for error here

File: test/e2e/framework/deploy/deploy.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	wd, err := os.Getwd()
+	Expect(err).To(BeNil())
+
+	providerResourceAbsolutePath, err := filepath.Abs(fmt.Sprintf("%s/../../%s", wd, providerResourceDirectory))
+	Expect(err).To(BeNil())
+
+	for _, resource := range providerResources {
+		exec.KubectlApply(kubeconfigPath, framework.NamespaceKubeSystem, []string{"-f", fmt.Sprintf("%s/%s", providerResourceAbsolutePath, resource)})
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:49]: We should check for error here

File: test/e2e/framework/deploy/deploy.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+func InstallManifest(kubeconfigPath string) {
+	for _, resource := range driverResources {
+		exec.KubectlApply(kubeconfigPath, framework.NamespaceKubeSystem, []string{"-f", fmt.Sprintf("%s/%s", driverResourcePath, resource)})
+	}
+
+	wd, err := os.Getwd()
+	Expect(err).To(BeNil())
+
+	providerResourceAbsolutePath, err := filepath.Abs(fmt.Sprintf("%s/../../%s", wd, providerResourceDirectory))
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go:45]: I went back and forth on this. Let me update to raw github url

File: test/e2e/e2e_suite_test.go (hunk 3 above)
Comment (nilekhc) [reply test/e2e/e2e_suite_test.go:52]: done

File: deployment/provider-azure-installer.yaml (hunk 2 above)
Comment (nilekhc) [reply deployment/provider-azure-installer.yaml:9]: Changes are required. Let me moved it to manifest_staging. This is also the reason to install provider resources from manifest_staging rather than from raw github source.

File: test/e2e/framework/deploy/deploy.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go:47]: reverted back to deployment from manifest staging.

File: .pipelines/templates/e2e-test-kind.yaml (hunk 1 above)
Comment (nilekhc) [reply .pipelines/templates/e2e-test-kind.yaml:8]: done

File: .pipelines/templates/e2e-test-kind.yaml [hunk 8]
```diff
@@ -1,37 +1,33 @@
 parameters:
   - name: driverWriteSecrets
     type: object
+  - name: helmTests
```
Comment (aramase) [inline .pipelines/templates/e2e-test-kind.yaml:4]: Is this still required?

File: .pipelines/templates/e2e-test-kind.yaml [hunk 9]
```diff
@@ -1,37 +1,33 @@
...
+    type: object
 
 jobs:
   - ${{ each driverWriteSecret in parameters.driverWriteSecrets }}:
     - job:
       displayName: ${{ format('e2e_test_kind/driver-write-secrets={0}', driverWriteSecret) }}
-      timeoutInMinutes: 10
+      timeoutInMinutes: 15
```
Comment (aramase) [inline .pipelines/templates/e2e-test-kind.yaml:11]: Do we need to increase the timeout? Are we seeing timeouts?

File: azure-pipelines.yml [hunk 10]
```diff
@@ -24,3 +24,6 @@ jobs:
       driverWriteSecrets:
       - "true"
       - "false"
+      helmTests:
```
Comment (aramase) [inline azure-pipelines.yml:27]: This is no longer required?

File: manifest_staging/deployment/provider-azure-installer.yaml [hunk 11]
```diff
@@ -9,6 +9,7 @@ metadata:
   labels:
     app: csi-secrets-store-provider-azure
   name: csi-secrets-store-provider-azure
+  namespace: kube-system
```
Comment (aramase) [inline manifest_staging/deployment/provider-azure-installer.yaml:12]: This is a breaking change. If the intention is to deploy the pods in `kube-system` then just `kubectl apply -n kube-system -f <file>` should be sufficient as the default namespace isn't set in the yamls.

File: manifest_staging/deployment/provider-azure-installer.yaml [hunk 12]
```diff
@@ -56,8 +57,13 @@ spec:
               drop:
               - ALL
           volumeMounts:
-            - mountPath: "/provider"
-              name: providervol
+            - name: provider-vol
+              mountPath: /provider
+            - name: mountpoint-dir
+              mountPath: /var/lib/kubelet/pods
+              mountPropagation: HostToContainer
```
Comment (aramase) [inline manifest_staging/deployment/provider-azure-installer.yaml:64]: Why are we adding this back? This is no longer required with the switch to driver writing secrets.

File: manifest_staging/deployment/provider-azure-installer.yaml [hunk 13]
```diff
@@ -56,8 +57,13 @@ spec:
...
+            - mountPath: /etc/kubernetes/custom_environment.json
```
Comment (aramase) [inline manifest_staging/deployment/provider-azure-installer.yaml:65]: This file is only used while testing so I don't think we should be actually adding it to the published yamls?

File: test/e2e/framework/deploy/deploy.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	Expect(err).To(BeNil())
+
+	for _, resource := range providerResources {
+		err := exec.KubectlApply(kubeconfigPath, framework.NamespaceKubeSystem, []string{"-f", fmt.Sprintf("%s/%s", providerResourceAbsolutePath, resource)})
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:51]: Are we not replacing the images in the yamls? If not, then the tests will fail when we make a change that needs a new image?

File: .pipelines/templates/e2e-test-kind.yaml (hunk 9 above)
Comment (nilekhc) [reply .pipelines/templates/e2e-test-kind.yaml:11]: It [failed](https://dev.azure.com/azure/secrets-store-csi-driver-provider-azure/_build/results?buildId=27078&view=logs&j=7f33e5bd-7764-5d8a-ba2e-506e078b9c3f) earlier. But let me revert it back and monitor since we created separate job for yaml testing.

File: deployment/provider-azure-installer.yaml (hunk 2 above)
Comment (nilekhc) [reply deployment/provider-azure-installer.yaml]: Reverted all changes and adding test specific changes during test resource install.

File: manifest_staging/deployment/provider-azure-installer.yaml (hunk 12 above)
Comment (nilekhc) [reply manifest_staging/deployment/provider-azure-installer.yaml]: reverted

File: test/e2e/framework/deploy/deploy.go (hunk 14 above)
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go]: Updating configuration with e2e image before deploy manifest testing.

File: .pipelines/templates/e2e-test-kind.yaml (hunk 9 above)
Comment (nilekhc) [reply .pipelines/templates/e2e-test-kind.yaml]: reverted back to 10m

File: manifest_staging/deployment/provider-azure-installer.yaml (hunk 13 above)
Comment (nilekhc) [reply manifest_staging/deployment/provider-azure-installer.yaml]: reverted

File: test/e2e/framework/config.go [hunk 15]
```diff
@@ -26,6 +26,7 @@ type Config struct {
 	PodIdentityUserAssignedIdentityID string `envconfig:"POD_IDENTITY_USER_ASSIGN_IDENTITY_ID"`
 	ResourceGroup                     string `envconfig:"RESOURCE_GROUP"`
 	IsUpgradeTest                     bool   `envconfig:"IS_UPGRADE_TEST"`
+	IsHelmTest                        bool   `envconfig:"IS_HELM_TEST" default:"true"`
```
Comment (aramase) [inline test/e2e/framework/config.go:29]: Could you add this to DeepCopy method too? `copy.IsHelmTest = c.IsHelmTest`

File: test/e2e/framework/deploy/deploy.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/secrets-store-csi-driver-provider-azure/test/e2e/framework"
+	"github.com/Azure/secrets-store-csi-driver-provider-azure/test/e2e/framework/exec"
+	appsv1 "k8s.io/api/apps/v1"
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:14]: Move the appsv1 import to next line. The imports are grouped 

```
import (
  standard packages

  internal packages

  external packages
)
```

File: test/e2e/framework/deploy/deploy.go [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/yaml"
+
+	. "github.com/onsi/gomega"
+)
+
+var (
+	driverResourcePath        = "https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/master/deploy"
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:22]: With the helm charts we test a specific version of the driver as defined in the `requirements`. Should we do the same here too and use a tagged release url instead of the `master` one?

File: test/e2e/framework/deploy/deploy.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
+// +build e2e
+
+package deploy
+
+import (
+	"fmt"
+	"io/ioutil"
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:7]: `io/ioutil` is deprecated in go 1.16: https://golang.org/doc/go1.16#ioutil. Let's use `os` package.

File: test/e2e/framework/deploy/deploy.go [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+	Expect(err).To(BeNil())
+
+	for _, resource := range providerResources {
+		file, err := os.Open(fmt.Sprintf("%s/%s", providerResourceAbsolutePath, resource))
+		Expect(err).To(BeNil())
+
+		fileBytes, err := ioutil.ReadAll(file)
+		Expect(err).To(BeNil())
+
+		// Extract DS yaml
```
Comment (aramase) [inline test/e2e/framework/deploy/deploy.go:61]: Could you more info in the comment on why we need to extract the daemonset yaml from the single yaml?

File: test/e2e/e2e_suite_test.go [hunk 20]
```diff
@@ -41,27 +42,35 @@ var _ = BeforeSuite(func() {
...
-	kubeClient = clusterProxy.GetClient()
-	kubeconfigPath = clusterProxy.GetKubeconfigPath()
+	if !config.IsHelmTest {
+		By("Installing Secrets Store CSI Driver and Azure Key Vault Provider via kubectl from deployment manifest.")
+		deploy.InstallManifest(kubeconfigPath, config)
+
+		return
+	}
+
+	if !helm.ReleaseExists() { //If helm release exists, it means either cluster upgrade test or backward compatibility test is underway.
```
[the remark this answers is not in the document]
Comment (nilekhc) [reply test/e2e/e2e_suite_test.go:62]: done

File: test/e2e/framework/config.go (hunk 15 above)
Comment (nilekhc) [reply test/e2e/framework/config.go:29]: done

File: test/e2e/framework/deploy/deploy.go (hunk 18 above)
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go:7]: removed

File: test/e2e/framework/deploy/deploy.go (hunk 19 above)
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go:61]: updated comment

File: test/e2e/framework/deploy/deploy.go (hunk 17 above)
Comment (nilekhc) [reply test/e2e/framework/deploy/deploy.go]: makes sense. Updated to `v0.1.0`

**Review by aramase — APPROVED** [review]
